lib/graphql/batch/query_group.rb

```diff
@@ -1,35 +1,48 @@
 module GraphQL::Batch
   class QueryGroup < QueryContainer
-    def initialize(queries, &block)
-      @pending_queries = queries.dup
-      @pending_queries.each do |query|
-        query.query_listener = self
+    def initialize(queries)
+      raise ArgumentError, "QueryGroup.new no longer takes a block, use #then instead" if block_given?
+      super(self)
+      @pending_queries = Set.new
+      @results = []
+
+      queries.each_with_index do |query, i|
+        if query.is_a?(QueryContainer)
+          @pending_queries.add(query)
+        else
+          @results[i] = query
+        end
+      end
+
+      if @pending_queries.empty?
+        complete(@results)
+      else
+        queries.each_with_index do |query, i|
+          next unless query.is_a?(QueryContainer)
+          query.then do |result|
+            @results[i] = result
+            @pending_queries.delete(query)
+            if @pending_queries.empty?
+              complete(@results)
+            end
+          end.rescue do |err|
+            unless completed?
```

Comment:
when could this be called and we're already completed? If we're completed, then all our children have returned results, and so none of them could raise the error?